### PR TITLE
osd: support loop devices for test clusters

### PR DESCRIPTION
**Pull request by peng225**
**Description of your changes:**

Now that `ceph-volume` supports loop devices, I believe Rook can also support them.
cf. https://github.com/ceph/ceph/pull/46375
cf. https://github.com/rook/rook/issues/7206

I added a new parameter `ROOK_CEPH_ALLOW_LOOP_DEVICES` to `rook-ceph-operator-config` ConfigMap.
Rook users can use loop devices by setting the parameter to `true`.

**Which issue is resolved by this Pull Request:**
Resolves no issue.

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [x] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [x] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.

### CODE REVIEW

File: pkg/clusterd/disk.go [hunk 1]
```diff
@@ -213,3 +215,11 @@ func PopulateDeviceUdevInfo(d string, executor exec.Executor, disk *sys.LocalDis
 
 	return disk, nil
 }
+
+func SetAllowLoopDevices(val bool) {
+	allowLoopDevices = val
```
[the remark this answers is not in the document]
Comment (peng225) [reply pkg/clusterd/disk.go:221]: Thank you for your comment.

`SetAllowLoopDevices` and `LoopDevicesAllowed` functions are used for two kinds of processes.
1. Operator
2. Pods (prepare job, discover daemonset, and cleanup job)

As for the second one, `SetAllowLoopDevices` was called at the beginning of each pod, and then `LoopDevicesAllowed` was used at `supportedDeviceType` to check if loop devices are allowed. I could eliminate the setter/getter call by introducing `GetAllowLoopDevices()` as you suggested.

However, for the operator, it should create each pod based on the latest `ROOK_CEPH_ALLOW_LOOP_DEVICES` parameter. So we need some means to know the value of it when creating these pods (at `pkg/operator/ceph/cluster/cleanup.go`, `pkg/operator/ceph/cluster/osd/provision_spec.go`, and `pkg/operator/discover/discover.go`).

What do you think is good for that? The following list is what I came up with.

- Use `SetAllowLoopDevices` and `LoopDevicesAllowed` as it is.
  - In that case, it will be better to move them to some other file since it is no longer used by `supportedDeviceType()`.
- Define a new environment variable for the operator process, and use it to propagate the current value of `ROOK_CEPH_ALLOW_LOOP_DEVICES` parameter.
- Do not check the value of `ROOK_CEPH_ALLOW_LOOP_DEVICES` parameter at `operator.reconcile()`. Instead, check the value directory from the `rook-ceph-operator-config` ConfigMap when creating each pod.

[the remark this answers is not in the document]
Comment (peng225) [reply pkg/clusterd/disk.go]: > Define a new environment variable for the operator process, and use it to propagate the current value of ROOK_CEPH_ALLOW_LOOP_DEVICES parameter.

For now, I adopted this approach and declared a new env `ROOK_CEPH_ALLOW_LOOP_DEVICES`, which has the same name as the new ConfigMap parameter.

File: deploy/charts/rook-ceph/values.yaml [hunk 2]
```diff
@@ -54,6 +54,9 @@ pspEnable: true
 # Set the priority class for the rook operator deployment if desired
 # priorityClassName: class
 
+# If true, loop devices are allowed to be used for osd
```
Comment (travisn) [inline deploy/charts/rook-ceph/values.yaml:57]: This is still a feature for a test cluster, correct? I just want to be clear we don't think it should be used in production. Same comment for everywhere else this is copied.
```suggestion
# If true, loop devices are allowed to be used for osds in test clusters
```

File: pkg/operator/ceph/controller.go [hunk 3]
```diff
@@ -134,6 +135,8 @@ func (r *ReconcileConfig) reconcile(request reconcile.Request) (reconcile.Result
 		return opcontroller.ImmediateRetryResult, err
 	}
 
+	reconcileAllowLoopDevices(r.config.Parameters)
```
Comment (travisn) [inline pkg/operator/ceph/controller.go:138]: Why is this needed? I could see that we would need to set the env var on the discovery daemon deployment spec, but not in the operator like this.

File: .github/workflows/canary-integration-test.yml [hunk 4]
```diff
@@ -1561,3 +1561,128 @@ jobs:
         if: failure() && env.USE_TMATE
         uses: mxschmitt/action-tmate@v3
         timeout-minutes: 60
+
+  loop-device:
```
Comment (travisn) [inline .github/workflows/canary-integration-test.yml:1565]: Instead of creating completely new canary integration tests for the loop devices, I wonder if we can just use the loop devices in an existing test? It seems like a very small difference in an existing test, instead of creating a new test for it.

File: deploy/charts/rook-ceph/values.yaml [hunk 5]
```diff
@@ -54,6 +54,9 @@ pspEnable: true
...
+allowLoopDevices: false
```
Comment (travisn) [inline deploy/charts/rook-ceph/values.yaml:58]: This setting needs to be documented in `Documentation/Helm-Charts/operator-chart.md`

File: pkg/operator/ceph/cluster/osd/provision_spec.go [hunk 6]
```diff
@@ -320,3 +325,11 @@ func (c *Cluster) provisionOSDContainer(osdProps osdProperties, copyBinariesMoun
 
 	return osdProvisionContainer, nil
 }
+
+func SetAllowLoopDevices(value string) {
```
Comment (travisn) [inline pkg/operator/ceph/cluster/osd/provision_spec.go:329]: Besides setting this on the pod specs, I don't think we should ever set this env var except during testing

**Review by travisn — CHANGES_REQUESTED** [review]

File: deploy/charts/rook-ceph/values.yaml (hunk 2 above)
Comment (peng225) [reply deploy/charts/rook-ceph/values.yaml]: I fixed it.

File: deploy/charts/rook-ceph/values.yaml (hunk 5 above)
Comment (peng225) [reply deploy/charts/rook-ceph/values.yaml:58]: I updated the document.

File: pkg/operator/ceph/controller.go (hunk 3 above)
Comment (peng225) [reply pkg/operator/ceph/controller.go:138]: This is related to https://github.com/rook/rook/pull/11268#discussion_r1014994697 .

For example, when the prepare job is created, we need to add an environment variable `CEPH_VOLUME_ALLOW_LOOP_DEVICES` to its pod definition **only when** loop devices are allowed by the new option `ROOK_CEPH_ALLOW_LOOP_DEVICES` in `rook-ceph-operator-config` ConfigMap. So `provisionOSDContainer` function needs to know the current value of `ROOK_CEPH_ALLOW_LOOP_DEVICES`.

In the current code, the `reconcile` function is the only place where the contents of the `rook-ceph-operator-config` ConfigMap is read. So I check the value of `ROOK_CEPH_ALLOW_LOOP_DEVICES` here and then use it when creating prepare pod. The new environment variable `ROOK_CEPH_ALLOW_LOOP_DEVICES` (which happens to have the same name as the new config) is just an intermediate variable.

Honestly, I feel it is clumsy. I will appreciate it if you have any ideas to improve it.

File: .github/workflows/canary-integration-test.yml (hunk 4 above)
Comment (peng225) [reply .github/workflows/canary-integration-test.yml]: I removed new tests and modified current tests to also handle loop devices.
I only modified `raw-disk` and `pvc` cases. If it is desirable to modify other tests, please let me know which test cases should be changed.

**Review by travisn — CHANGES_REQUESTED** [review]

File: deploy/charts/rook-ceph/values.yaml (hunk 2 above)
Comment (peng225) [reply deploy/charts/rook-ceph/values.yaml]: I also changed the commit and PR title to say that this feature is for test purposes only.

File: pkg/operator/ceph/controller.go (hunk 3 above)
Comment (peng225) [reply pkg/operator/ceph/controller.go]: Thank you for your comment.

I followed the pattern of `SetCephCommandsTimeout()` method and changed the code to use package variable again.

> But I'd rather not have a "reconcile" helper method if we're just setting a variable, so maybe call this directly:

I deleted `reconcileAllowLoopDevices` method.

File: pkg/operator/ceph/cluster/osd/provision_spec.go (hunk 6 above)
Comment (peng225) [reply pkg/operator/ceph/cluster/osd/provision_spec.go]: I believe this is resolved in the discussion at https://github.com/rook/rook/pull/11268#discussion_r1017495887 .

**Review by travisn — APPROVED** [review]
LGTM! This can also enable us in a future PR to test with more OSDs where we have been limited in the past for different replication, crush configurations, etc.
